### DISCUSSION TITLE
Private keys and secure element support

### DISCUSSION THREAD

**Discussion by layton-dev** [strongswan/strongswan#1457] [Q&A] [answered]
Hello,

I'm trying to use a secure element (an ATECC608A) with Strongswan, to store a private key that is associated with my certificate.
I'm able to communicate with the secure element with the PKCS11 library, and I've configured the PKCS11 plugin for Strongswan (as I'm aware SmartCard authentication is supported).

But it seems Strongswan isn't able to communicate with the Secure Element; I was wondering what ipsec.secrets configuration I should use for this particular setup ?

Thank you !

**Comment by tobiasbrunner** [comment, marked as answer]
> But it seems Strongswan isn't able to communicate with the Secure Element

Why? Please provide more information (see #196 for log settings).

> I was wondering what ipsec.secrets configuration I should use for this particular setup?

Did you read [the docs](https://wiki.strongswan.org/projects/strongswan/wiki/PinSecret)? Also, you should rather use [swanctl.conf](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html#_secrets_tokensuffix) anyway. Either way, it's dependent on the contents of your SE than a particular setup. So please provide more information about that (e.g. listings from `pkcs11-tool`).

**Reply by layton-dev** [reply]
Hello,

Thank you for your quick reply ! I thought ipsec.secrets was still required, even after some research on my part in the docs.
I followed your advice and switched to `swanctl.conf`. Here is more detailed information about the setup :

### pkcs11-tool -L
```
Available slots:

Slot 0 (0x0): 0_3_0
  token label        : 00ABC
  token manufacturer : Microchip Technology Inc
  token model        : ATECC608A
  token flags        : rng, token initialized
  hardware version   : 0.2
  firmware version   : 255.255
  serial num         : ...
  pin min/max        : 0/128
```

### swanctl.conf

```
secrets {
        token0{
                handle = "0"
                module = "/usr/lib/aarch64-linux-gnu/libcryptoauth.so"
                pin = ...               
        }
}
```

### swanctl --load-all

```
loading 'token0' failed: loading private key from token failed
no authorities found, 0 unloaded
no pools found, 0 unloaded
```

### /var/log/syslog

```
Dec 10 16:49:43 debian charon: 16[CFG] no PKCS#11 module found having a keyid 00
Dec 10 16:49:43 debian charon: 16[LIB] no TPM 2.0 found
Dec 10 16:49:43 debian charon: 16[LIB] building CRED_PRIVATE_KEY - ANY failed, tried 9 builders
Dec 10 16:49:43 debian charon: 11[CFG] algorithm 'modp2048!' not recognized
Dec 10 16:49:43 debian charon: 09[CFG] algorithm 'modp2048!' not recognized 
```

It seems that Strongswan isn't able to detect the private key and / or slot inside the SE. I've checked that the certificate and public key loaded in Strongswan do correctly match the private key.

Thank you for your help,

**Reply by tobiasbrunner** [reply]
The `handle` should be the CKA_ID of the private key (check with `pkcs11-tool -O`). However, is the module actually loaded? Note that `module` in the token section doesn't reference a shared library but the (optional) name of a module section defined in [strongswan.conf](https://docs.strongswan.org/docs/5.9/config/strongswanConf.html#_charon_plugins_pkcs11) (that's where you specify the path to the shared library).

**Reply by layton-dev** [reply]
Hello, thank you for your answer and sorry for the delay, I've managed to track down the problem : the library I was using to communicate with the ATECC608A (cryptoauthlib) had no CKA_ID support until pretty recently. Updating the library did the trick, following your advice I've changed the handle in swanctl.conf to the private key's CKA_ID, and now I can successfully load the private key in strongswan !

Thanks again for your help !
